### ISSUE TITLE
(aws-logs): log retention update can race against the target lambda's log group creation resulting in OperationAbortedException

### ISSUE THREAD

**Issue by steved** [aws/aws-cdk#15709]
When a lambda runs a log group and stream are created if they do not already exist. When creating a CDK Lambda function, the log retention update can then race with the background log group creation resulting in an `OperationAbortedException` error.

### Reproduction Steps

Hard to reproduce reliably as it relies on a race condition between AWS and the log retention lambda.

### What did you expect to happen?

Log retention policy is properly applied.

### What actually happened?

```
2021-07-22T02:52:24.985Z	a53fc34d-6d93-4590-94fd-8115d0b93e3b	INFO	OperationAbortedException: A conflicting operation is currently in progress against this resource. Please try again.
    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:52:27)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:690:12)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:116:18) {
  code: 'OperationAbortedException',
  time: 2021-07-22T02:52:24.925Z,
  requestId: '[redacted]',
  statusCode: 400,
  retryable: false,
  retryDelay: 56.75647811158137
}
```

### Environment

  - **CDK CLI Version  :** 1.110.1
  - **Framework Version:** 1.110.1
  - **Node.js Version:** 10.24.0
  - **OS               :** Buster (container)
  - **Language (Version):** all, Python (3.8)
 
### Other

AWS log group creation cloudtrail event:
```
{
    "eventVersion": "1.08",
    "userIdentity": {
        "type": "AssumedRole",
        "principalId": "[...]",
        "arn": "[...]",
        "accountId": "[...]",
        "accessKeyId": "[...]",
        "sessionContext": {
            "sessionIssuer": {
               ...
            },
            "webIdFederationData": {},
            "attributes": {
                "creationDate": "2021-07-22T02:52:23Z",
                "mfaAuthenticated": "false"
            }
        }
    },
    "eventTime": "2021-07-22T02:52:24Z",
    "eventSource": "logs.amazonaws.com",
    "eventName": "CreateLogGroup",
    "awsRegion": "us-west-2",
    "sourceIPAddress": "[...]",
    "userAgent": "awslambda-worker/1.0  rusoto/0.42.0 rust/1.52.1 linux",
    "requestParameters": {
        "logGroupName": "/aws/lambda/mylambda"
    },
    "responseElements": null,
    "requestID": "[...]",
    "eventID": "[...]",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "apiVersion": "20140328",
    "managementEvent": true,
    "recipientAccountId": "...",
    "eventCategory": "Management"
}
```

Log retention API call:
```
{
    "eventVersion": "1.08",
    "userIdentity": {
        "type": "AssumedRole",
        "principalId": "...",
        "arn": "...",
        "accountId": "...",
        "accessKeyId": "...",
        "sessionContext": {
            "sessionIssuer": {
                ...
            },
            "webIdFederationData": {},
            "attributes": {
                "creationDate": "2021-07-22T02:52:23Z",
                "mfaAuthenticated": "false"
            }
        }
    },
    "eventTime": "2021-07-22T02:52:24Z",
    "eventSource": "logs.amazonaws.com",
    "eventName": "CreateLogGroup",
    "awsRegion": "us-west-2",
    "sourceIPAddress": "[...]",
    "userAgent": "aws-sdk-nodejs/2.880.0 linux/v12.22.1 exec-env/AWS_Lambda_nodejs12.x promise",
    "errorCode": "OperationAbortedException",
    "errorMessage": "A conflicting operation is currently in progress against this resource. Please try again.",
    "requestParameters": {
        "logGroupName": "/aws/lambda/mylambda"
    },
    "responseElements": null,
    "requestID": "[...]",
    "eventID": "[...]",
    "readOnly": false,
    "eventType": "AwsApiCall",
    "apiVersion": "20140328",
    "managementEvent": true,
    "recipientAccountId": "[...]",
    "eventCategory": "Management"
}
```

https://github.com/aws/aws-cdk/pull/2237 seems to be a fix for the same potential error on the actual log retention lambda's log group, but not the target.

**Comment by rix0rrr** [comment]
Easiest way to fix this is to update the Lambda to be resilient against this occurring.

**Comment by rix0rrr** [comment]
Easiest way to fix this is to update the Lambda to be resilient against this occurring.

**Comment by steved** [comment]
How would you recommend to approach this? Would the addition of the error code to this exception help? Or a global retry?

https://github.com/aws/aws-cdk/blob/2c4ef0131893e77d373c52b41c62d31847023446/packages/%40aws-cdk/aws-logs/lib/log-retention-provider/index.ts#L25

**Comment by rix0rrr** [comment]
I woukd say retry on the specific error

**Comment by ddl-denis-parnovskiy** [comment]
Please, could someone review the proposed fix?

We have 2 bugs in packages/@aws-cdk/aws-logs/lib/log-retention-provider/index.ts

1. It handles OperationAbortedException race for its own lambda but not for the lambda created elsewhere in CDK (this issue essential). 
2. When it handles OperationAbortedException race for its own lambda it skips retention assignment altogether while it should just ignore the log group creation error.

**Comment by slavman** [comment]
we still have this error when add logRetention. cdk version is 1.124.0

**Comment by m90** [comment]
I am also experiencing this issue when using `1.127.0` throughout.

**Comment by horsmand** [comment]
My team is hitting this with version `1.129.0`.

We have some bash scripting that allows us to deploy multiple CDK apps in parallel that share some base infrastructure and we are very consistently hitting this issue.

It actually seems worse with the new fix because before it would fail silently and let the deployment continue, but now it fails the whole CDK deployment due to the Lambda function now throwing errors instead of swallowing them.

I haven't captured logs from the Lambda yet, but I am seeing failures like this in the `cdk deploy` output: 
```
SingletonLambda2e19e24316ee4d1aa3c918d35eddd446/LogRetention (SingletonLambda2e19e24316ee4d1aa3c918d35eddd446LogRetention970D13ED) Received response status [FAILED] from custom resource. Message returned: A conflicting operation is currently in progress against this resource. Please try again.
```

This leads me to believe the retry logic isn't being hit at all, since the code for that looks like it should throw a new error with the message `Out of attempts to create a logGroup`.

**Comment by mmeylan** [comment]
We're also getting this error with `1.128.0` and `1.131.0`. It happens when trying to access a function's log group in order to setup an alarm based on errors in the logs. Here's the stack we're getting:

9:40:23 PM | CREATE_FAILED        | Custom::LogRetention                      | BgoSendgridEventWe...gRetention2CACE5F0
Received response status [FAILED] from custom resource. Message returned: A conflicting operation is currently in progress
against this resource. Please try again. (RequestId: 6e534485-2033-4eaa-a4fe-2c0cde8dbcf3)

```
	9:40:23 PM | CREATE_FAILED        | Custom::LogRetention                      | XXX...gRetention2CACE5F0
Received response status [FAILED] from custom resource. Message returned: A conflicting operation is currently in progress
against this resource. Please try again. (RequestId: 6e534485-2033-4eaa-a4fe-2c0cde8dbcf3)

	new LogRetention (/local/node_modules/@aws-cdk/aws-logs/lib/log-retention.ts:53:22)
	\_ Function.get logGroup [as logGroup] (/local/node_modules@aws-cdk/aws-lambda/lib/functiod response status [FAILED] from custom resource. Message returned: A conflicting operation is currently in progress
n.ts:550:28)
	\_ LambdaMonitoring.setupAlarm (/local/cdk/lib/monitoring/lambda-monitoring.ts:83:35)eceived response status [FAILED] from custom resource. Message returned: A conflicting operation is currently in progress
	\_ new LambdaMonitoring (/local/cdk/lib/monitoring/lambda-monitoring.ts:27:24)
	\_ BackendStack.setupLambdaMonitoring (/local/cdk/lib/backend-stack.ts:1616:7)
	\_ BackendStack.setupCloudwatchDashboards (/local/cdk/lib/backend-stack.ts:1436:10)
	\_ BackendStack.setupCloudwatch (/local/cdk/lib/backend-stack.ts:1380:10)
	\_ new BackendStack (/local/cdk/lib/backend-stack.ts:78:12)
	\_ StackBuilder.buildStack (/local/cdk/lib/stack-builder.ts:220:28)
	\_ StackBuilder.build (/local/cdk/lib/stack-builder.ts:18:30)
	\_ processTicksAndRejections (internal/process/task_queues.js:95:5)
```

**Comment by oscarwest** [comment]
We've been hitting this as well

**Comment by horsmand** [comment]
I've opened a new ticket since many people have been reporting the issue seems to be happening still. I believe I've made some findings in regards to why it might still be causing problems: https://github.com/aws/aws-cdk/issues/17546
